lib/client/db/postgres/repl/repl.go

```diff
@@ -0,0 +1,288 @@
+	config, err := pgconn.ParseConfig(fmt.Sprintf("postgres://%s", hostnamePlaceholder))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	config.User = route.Username
+	config.Database = route.Database
+	config.ConnectTimeout = defaults.DatabaseConnectTimeout
+	config.RuntimeParams = map[string]string{
+		applicationNameParamName: applicationNameParamValue,
+	}
+	config.TLSConfig = nil
+
+	// Provide a lookup function to avoid having the hostname placeholder to
+	// resolve into something else. Note that the returned value won't be used.
+	config.LookupFunc = func(_ context.Context, _ string) ([]string, error) {
+		return []string{hostnamePlaceholder}, nil
+	}
+	config.DialFunc = func(_ context.Context, _, _ string) (net.Conn, error) {
+		return serverConn, nil
+	}
+
+	return &REPL{
+		connConfig: config,
+		client:     client,
+		serverConn: serverConn,
+		route:      route,
+		term:       term.NewTerminal(client, ""),
+		commands:   initCommands(),
+	}, nil
+}
+
+// Run starts and run the PostgreSQL REPL session. The provided context is used
+// to interrupt the execution and clean up resources.
+func (r *REPL) Run(ctx context.Context) error {
+	pgConn, err := pgconn.ConnectConfig(ctx, r.connConfig)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer pgConn.Close(context.TODO())
+
+	// term.Terminal blocks reads/writes without respecting the context. The
+	// only thing that unblocks it is closing the underlaying connection (in
+	// our case r.client). On this goroutine we only watch for context
+	// cancelation and close the connection. This will unblocks all terminal
+	// reads/writes.
+	ctxCancelCh := make(chan struct{})
+	defer close(ctxCancelCh)
+	go func() {
+		select {
+		case <-ctx.Done():
+			_ = r.client.Close()
+		case <-ctxCancelCh:
+		}
+	}()
+
+	if err := r.presentBanner(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	var (
+		multilineAcc     strings.Builder
+		readingMultiline bool
+	)
+
+	lead := lineLeading(r.route)
+	leadSpacing := strings.Repeat(" ", len(lead))
+	r.term.SetPrompt(lineBreak + lead)
+
+	for {
+		line, err := r.term.ReadLine()
+		if err != nil {
+			return trace.Wrap(formatTermError(ctx, err))
+		}
+
+		// ReadLine should always return the line without trailing line breaks,
+		// but we still require to remove trailing and leading spaces.
+		line = strings.TrimSpace(line)
+
+		var reply string
+		switch {
+		case strings.HasPrefix(line, commandPrefix) && !readingMultiline:
+			var exit bool
+			reply, exit = r.processCommand(line)
+			if exit {
+				return nil
+			}
+		case strings.HasSuffix(line, executionRequestSuffix):
+			var query string
+			if readingMultiline {
+				multilineAcc.WriteString(lineBreak + line)
+				query = multilineAcc.String()
+			} else {
+				query = line
+			}
+
+			// Reset multiline state.
+			multilineAcc.Reset()
+			readingMultiline = false
+			r.term.SetPrompt(lineBreak + lead)
+
+			reply = formatResult(pgConn.Exec(ctx, query).ReadAll()) + lineBreak
```

Comment:
What would be the result with multiple `;` in a single line? E.g.:
```select * from
foo; select * from bar;
```

I'm guessing this will simply run multiple queries?

Reply:
Yes, but this is up to `pgconn` to solve.